test-data/unit/check-based-union-join.test

```diff
@@ -0,0 +1,62 @@
+[case testSimpleUnionAtJoin]
+a = [1, ""]
+reveal_type(a)  # N: Revealed type is "list[int | str]"
+
+
+[case testSubtype]
+class A: pass
+class B(A): pass
+
+reveal_type([A(), B()])  # N: Revealed type is "list[__main__.A]"
+
+
+[case testConditionalExpression]
+class A: pass
+class B: pass
+
+res = A() if bool() else B()
+
+reveal_type(res)  # N: Revealed type is "__main__.A | __main__.B"
+
+
+[case testTupleUnion]
+reveal_type([(1, ""), (None, False)])  # N: Revealed type is "list[(int, str) | (None, bool)]"
+[builtins fixtures/tuple.pyi]
+
+
+[case testUnionType]
+reveal_type(type([1, ""][0]))  # N: Revealed type is "type[int] | type[str]"
+
+
+[case testAbstractTypeCollection]
+from typing import Protocol
+class Base(Protocol):
+    a: int
+class A(Base):
+    pass
+class B(Base):
+    pass
+class C(Base):
+    pass
```

Comment:
these protocols are all structurally the same. did you mean to add a property to each of them

Reply:
They aren't protocols, they implement the interface of a protocol.